File: packages/agent/agent/subagents/extractor/tools/save_extraction.ts
import { randomUUID } from "node:crypto";

import { getDb } from "@clara-financas/db";
import { extractionStagings } from "@clara-financas/db/schema/extraction-staging";
import { documents } from "@clara-financas/db/schema/ledger";
import { forTenant } from "@clara-financas/db/tenant-scope";
import { ExtractionResultSchema } from "@clara-financas/views/agent-contracts";
import { and, eq } from "drizzle-orm";
import { defineTool } from "eve/tools";

import { notFound } from "../../../lib/errors";
import { requireTenantCaller } from "../../../lib/tenant";

/**
 * Persiste a extração completa e devolve só o recibo.
 *
 * É a metade do extrator na passagem por REFERÊNCIA: as 100+ linhas lidas do
 * PDF não voltam mais pelo contexto do coordenador — ficam na staging, e o
 * recibo carrega o `extractionId` que `propose_batch_from_extraction` consome.
 *
 * Sobre o isolamento: a invariante do extrator sempre foi "não alcança o
 * razão", e ela continua de pé — esta tool escreve numa tabela de staging
 * descartável, sem grant algum sobre `batches`/`transactions`. O que mudou é
 * que "nenhuma tool de escrita" virou "nenhuma escrita no razão".
 */
export default defineTool({
  description:
    "Persists the complete extraction of a document and returns an extractionId. Call EXACTLY ONCE, with every transaction you read, right before returning your structured result. Your final output must be the receipt — the coordinator proposes the batch from the extractionId, so never retype the transactions anywhere else.",
  inputSchema: ExtractionResultSchema,
  async execute(input, ctx) {
    const { tenantId } = requireTenantCaller(ctx);
    const db = getDb();

    // Revalidação explícita: o inputSchema já validou a forma, o parse aplica
    // os defaults (warnings) e garante que o payload persistido é exatamente o
    // contrato — não o que o transporte deixou passar.
    const payload = ExtractionResultSchema.parse(input);
    const extractionId = `ext_${randomUUID().replace(/-/g, "").slice(0, 20)}`;
    // A sessão do PAI, quando esta tool roda dentro de um subagente delegado:
    // chave de correlação que nenhum modelo digita.
    const parentSessionId = ctx.session.parent?.sessionId ?? null;

    const result = await forTenant(
      tenantId,
      async (tx) => {
        const [document] = await tx
          .select({ id: documents.id })
          .from(documents)
          .where(and(eq(documents.id, payload.documentId), eq(documents.tenantId, tenantId)))
          .limit(1);

        if (document === undefined) {
          return notFound(
            "documento_nao_encontrado",
            `Nenhum documento com o id ${payload.documentId}.`,
            { hint: "Use o documentId exatamente como veio na requisição do coordenador." },
          );
        }

        // Uma extração por documento: refazer a leitura substitui a anterior,
        // no mesmo espírito da idempotência de proposta por documento.
        await tx
          .delete(extractionStagings)
          .where(
            and(
              eq(extractionStagings.tenantId, tenantId),
              eq(extractionStagings.documentId, payload.documentId),
            ),
          );

        await tx.insert(extractionStagings).values({
          id: extractionId,
          tenantId,
          documentId: payload.documentId,
          parentSessionId,
          payload,
        });

        return { ok: true as const };
      },
      db,
    );

    if ("error" in result) return result;

    return {
      extractionId,
      documentId: payload.documentId,
      transactionCount: payload.transactions.length,
      note: "Extração guardada. Devolva o recibo com este extractionId — não repita as transações no texto.",
    };
  },
});

File: services/agent/clara/tools/extractor/save_extraction.py
"""Persiste a extração completa e devolve só o recibo — porta de
`subagents/extractor/tools/save_extraction.ts`.

É a metade do extrator na passagem por REFERÊNCIA: as 100+ linhas lidas do
PDF não voltam pelo contexto da coordenadora — ficam na staging, e o recibo
carrega o `extraction_id` que `propose_batch_from_extraction` consome.

Isolamento: a invariante do extrator é "não alcança o razão". Esta função só
escreve em `extraction_stagings`, tabela descartável sem relação com
`batches`/`transactions`.
"""

from __future__ import annotations

import uuid

from sqlalchemy import delete, select
from sqlalchemy.orm import Session

from clara.db.models import Document, ExtractionStaging
from clara.tools.errors import ToolError, not_found
from clara.views.agent_contracts import ExtractionReceipt, ExtractionResult


def save_extraction(
    session: Session,
    tenant_id: str,
    payload: ExtractionResult,
    *,
    parent_session_id: str | None = None,
) -> ExtractionReceipt | ToolError:
    document = session.execute(
        select(Document.id).where(
            Document.id == payload.document_id, Document.tenant_id == tenant_id
        )
    ).scalar_one_or_none()

    if document is None:
        return not_found(
            "documento_nao_encontrado", f"Nenhum documento com o id {payload.document_id}.",
            hint="Use o document_id exatamente como veio na requisição da coordenadora.",
        )

    extraction_id = f"ext_{uuid.uuid4().hex[:20]}"

    # Uma extração por documento: refazer a leitura substitui a anterior.
    session.execute(
        delete(ExtractionStaging).where(
            ExtractionStaging.tenant_id == tenant_id,
            ExtractionStaging.document_id == payload.document_id,
        )
    )
    session.add(
        ExtractionStaging(
            id=extraction_id, tenant_id=tenant_id, document_id=payload.document_id,
            parent_session_id=parent_session_id,
            payload=payload.model_dump(mode="json"),
        )
    )

    document_row = session.get(Document, payload.document_id)
    if document_row is not None:
        document_row.kind = payload.document_kind
    session.flush()

    return ExtractionReceipt(
        extraction_id=extraction_id, document_id=payload.document_id,
        document_kind=payload.document_kind, opening_balance=payload.opening_balance,
        closing_balance=payload.closing_balance, issuer=payload.issuer,
        period_start=payload.period_start, period_end=payload.period_end, due_date=payload.due_date,
        declared_total=payload.declared_total,
        declared_subtotals=payload.declared_subtotals,
        transaction_count=len(payload.transactions), warnings=payload.warnings,
    )
